mne/channels/channels.py

```diff
@@ -987,6 +989,118 @@ def add_channels(self, add_list, force_update_info=False):
+
+        Parameters
+        ----------
+        group_by : dict
+            Specifies which channels are aggregated into a single channel, with
+            aggregation method determined by the ``method`` parameter. One new
+            pseudo-channel is made per dict entry; the dict values must be
+            lists of picks (integer indices of ``ch_names``). For example::
+
+                group_by=dict(Left_ROI=[1, 2, 3, 4], Right_ROI=[5, 6, 7, 8])
+
+            Note that within a dict entry all channels must have the same type.
+        %(method)s
+            If callable, the callable must accept one positional input (data of
+            shape ``(n_channels, n_times)``, ``(n_epochs, n_channels,
+            n_times)``, or ``(n_evokeds, n_channels, n_time)``) and return an
+            :class:`array <numpy.ndarray>` of shape ``(n_times)``, ``(n_epochs,
+            n_times)``, or ``(n_evokeds, n_times)``. For example with an
+            instance of Raw::
+
+                method = lambda data: np.median(data, axis=0)
+
+            Defaults to ``mean``.
+
+        Returns
+        -------
+        inst : instance of Raw, Epochs, or Evoked
+            A new instance.
+        """
+        from ..io import BaseRaw, RawArray
+        from ..epochs import BaseEpochs
+        from ..evoked import Evoked
+        from .. import EpochsArray, EvokedArray
+
+        # Check to see if data is preloaded
+        _check_preload(self, 'Combining channels')
+
+        ch_axis = len(self._data.shape) - 2  # idx of the channel axis
+        ch_idx = _picks_to_idx(self.info, None)
+        ch_types = _get_channel_types(self.info)
+        group_by = deepcopy(group_by)
+
+        # Convert string values of ``method`` into callables
+        if isinstance(method, str):
+            method_dict = {key: partial(getattr(np, key), axis=ch_axis)
+                           for key in ('mean', 'median', 'std')}
+            try:
+                method = method_dict[method]
+            except KeyError:
+                raise ValueError('"method" must be a callable, or one of '
+                                 '"mean", "median", or "std"; got {}'
+                                 ''.format(method))
+
+        # Check correctness of combinations
+        for this_group, these_picks in group_by.items():
+            # Check if channel indices are out of bounds
+            if not all(idx in ch_idx for idx in these_picks):
+                raise ValueError('Some channel indices are out of bounds.')
+            # Check if heterogeneous sensor type combinations
+            this_ch_type = np.array(ch_types)[these_picks]
+            if len(set(this_ch_type)) > 1:
+                types = ', '.join(set(this_ch_type))
+                raise ValueError('Cannot combine sensors of different types; '
+                                 '"{}" contains types {}.'
+                                 ''.format(this_group, types))
```

Comment:
since our minimum supported Python version recently changed to 3.6, we can use [f-strings](https://docs.python.org/3/reference/lexical_analysis.html#f-strings) instead of `str.format()`.  Our plan is to use f-strings in new code, and gradually update old code to use f-strings as it gets edited. These lines would become:

```suggestion
                raise ValueError('Cannot combine sensors of different types; '
                                 f'"{this_group}" contains types {types}.')
```